[No QA] Add Pusher Automated Tests

marcaaron opened the pull request:
### Details
Second attempt to make Pusher tests. Closed the first PR here https://github.com/Expensify/Expensify.cash/pull/771

### Fixed Issues
Fixes https://github.com/Expensify/Expensify/issues/145497

### Tests
Automated

### Tested On
❌ 

### Screenshots
❌

marcaaron committed:
change name

marcaaron committed:
fix created being in the wrong order

marcaaron committed:
fix style

marcaaron committed:
fix style then get rid of async

marcaaron committed:
fix style

marcaaron committed:
Merge remote-tracking branch 'origin' into marcaaron-pusher-tests

marcaaron committed:
fix tests to use Onyx

marcaaron committed:
fix conflicts

marcaaron committed:
fix API.js

marcaaron committed:
tests broke somehow

marcaaron committed:
fix conflicts

marcaaron committed:
Fix this test up

marcaaron committed:
remove some things

marcaaron committed:
add comment

marcaaron committed:
fix conflicts

AndrewGable commented on src/libs/actions/Report.js:
```diff
@@ -287,7 +288,15 @@ function subscribeToReportCommentEvents() {
+    // Note: When running automated tests the act of calling Pusher.isSubscribed will create a mock
+    // channel already in a subscribed state. So, we must skip this check when tests are running or else it
+    // will prevent us from binding an event to the channel below.
+    if (!CONFIG.IS_JEST_RUNNING && pusherSubscribedOrSubscribing) {
```
Is there any way we could mock `subscribeToReportCommentEvents` so we don't have to have this check?

marcaaron replied:
So, the way I am testing this now is a sort of verification that the `subscribeToReportCommentEvents()` action is doing what it's supposed to do. Which is, subscribe to the channel and handle the payload.

One think we could do is add a `force` parameter to the action that skips this check and always tries to subscribe to Pusher no matter what. Then we could lose this config. 

But I'd be curious to hear if you have any other ideas on how to handle this!

marcaaron replied:
also very curious to learn more about why if it's the case that you are not loving having this be a `CONFIG`.

marcaaron replied:
Wait! I've got an idea. We can mock those Pusher methods so they just return false no matter what.

marcaaron committed:
fix conflicts

marcaaron committed:
fix ua mocks

marcaaron committed:
remove uneeded

marcaaron committed:
Mock Pusher methods instead of adding the IS_JEST_RUNNING

marcaaron committed:
remove CONFIG dependency

marcaaron committed:
fix comment

AndrewGable reviewed (dismissed):
Looking a whole lot better! Adding @tgolen for review since he had comments on the last PR.

tgolen commented on tests/actions/ReportTest.js:
```diff
@@ -0,0 +1,73 @@
+        });
+
+        // Set up Onyx with some test user data
+        Onyx.set(ONYXKEYS.SESSION, {accountID: 1, email: 'test@test.com'});
+        return waitForPromisesToResolve()
```
I still haven't quite gotten the hang of these, so forgive me if this isn't how it's meant to be done... is the test that has been written any different than doing this?

```
return Onyx.set(ONYXKEYS.SESSION, {accountID: 1, email: 'test@test.com'})
```

My thinking is just that it's odd to see `waitForPromisesToResolve()` being used when the promise returned from `Onyx.set()` should resolve and work as well (and it's more obvious what is happening because the test uses the same promise chain).

marcaaron replied:
That makes sense! And I wouldn't mind updating it, but want to walk you through my thought process here to see if you have similar thoughts...

We could wait for `Onyx.set()` instead of doing `waitForPromisesToResolve()`, but I was thinking that for consistency we might want to default to `waitForPromisesToResolve()`. Even though unnecessary it might make it easier to understand how these tests are structured and deal with async code that doesn't necessarily have an "await-able" promise. Maybe this practice actually makes it harder to tell that's what `waitForPromisesToResolve()` is for?

Another thought I keep having is that if "everything async returned a `Promise`" (or even just the async "actions") then the test code might be more straight-forward and it would be more obvious to tell what exactly we are waiting for.

marcaaron replied:
I'll change this for now since I think I did have it this way at one point.

tgolen replied:
OK, you brought up some really good points. What I'd like to avoid is creating a situation where people get too promise-happy and just return promises from everything. That essentially promotes a ton of promise chains, and that was what plagued the application from an early age. I had to do a lot of work to remove all those promise chains and I'd hate to go back down that path. So, now that I think about it, I think your original solution here is actually superior.

tgolen commented on tests/actions/ReportTest.js:
```diff
@@ -0,0 +1,73 @@
+        // Set up Onyx with some test user data
+        Onyx.set(ONYXKEYS.SESSION, {accountID: 1, email: 'test@test.com'});
+        return waitForPromisesToResolve()
+            .then(() => {
+                subscribeToReportCommentEvents();
+                return waitForPromisesToResolve();
+            })
```
Same here. Is this the same?

```
.then(subscribeToReportCommentEvents)
```

marcaaron replied:
`subscribeToReportCommentEvents()` is a good example of something that doesn't return a `Promise` but does async work in the call to `Pusher.subscribe()`. So, it's not really the same since we want to wait until we are subscribed.

tgolen replied:
Ah, I didn't realize it didn't return a promise. Let's leave it the way it is.

tgolen commented on tests/actions/ReportTest.js:
```diff
@@ -0,0 +1,73 @@
+
+                // Once this happens we should see the comment get processed by the callback and added to the
+                // storage so we must wait for promises to resolve again and then verify the data is in Onyx.
+                return waitForPromisesToResolve();
```
I can see how maybe this is necessary to prevent any race conditions with the local data stored in `reportActions` not being up-to-date with Onyx yet.

marcaaron replied:
That's right. The callback should happen in realtime and then we set the data async via `Onyx`.

tgolen commented on tests/actions/ReportTest.js:
```diff
@@ -0,0 +1,73 @@
+import waitForPromisesToResolve from '../utils/waitForPromisesToResolve';
+
+describe('actions/Report', () => {
+    it('should subscribe to the private-user-accountID channel and handle reportComment events', () => {
```
I think this description can be improved to explain "why" this is being tested. Specifically, I think it's testing that when a report action is added, then the resulting Pusher event results in the new report action being put into Onyx. The description of which channel is being subscribed to or the name of the event being handled is secondary information.

marcaaron replied:
Yeah I agree that this could use some more emphasis on the relevant info.

tgolen reviewed (changes requested):
This looks like a good use for the tests, but I'm concerned that it's only testing half the flow. What about including more of the flow such as, using `Report.addAction()`, testing that an optimistic action is created in Onyx, then the pusher event happens, and verify that the optimistic action was replaced by the real action?

This comment is more of a NAB, but I also really like when tests conform to the format:
```
// Given <initial_condition>
...  code that sets initial condition
// When <something_happens>
... code that does something
// Then <expectation>
... code that performs the assertion
```

I know we use this pattern a lot in Chronos and I think it's really great because it provides lots of context for someone else that has to come in to maintain the test and understand the "why" behind everything that was done. Since we only have a few tests written so far, I think it would be a good time to adopt that same format, but if you are not sold on it, I can pull it out into a separate discussion.

marcaaron committed:
Merge remote-tracking branch 'origin' into marcaaron-pusher-tests

marcaaron committed:
Fix up comment. Use Onyx.set instead of waitForPromisesToResolve()

marcaaron committed:
Improve tests. Add TestHelper.

marcaaron committed:
fix lint

marcaaron commented:
Alright, so I started to write a much better test for `Report.addAction()`, but it was so much better than it ended up revealing a bug in `react-native-onyx`. Fixed the bug over there and also cleaned up the main tests in that repo.

https://github.com/Expensify/react-native-onyx/pull/36/files

>I know we use this pattern a lot in Chronos and I think it's really great because it provides lots of context for someone else that has to come in to maintain the test and understand the "why" behind everything that was done

Yeah that does sound pretty organized. I'll check out the Chronos tests and investigate.

tgolen commented on tests/actions/ReportTest.js:
```diff
@@ -0,0 +1,102 @@
+
+        // When using the Pusher mock the act of calling Pusher.isSubscribed will create a
+        // channel already in a subscribed state. These methods are normally used to prevent
+        // duplicated subscriptions, but we don't need them for this test.
```
```suggestion
        // duplicated subscriptions, but we don't need them for this test so forcing them to return false will make the testing less complex.
```
This adds a little more clarity.

tgolen commented on tests/actions/ReportTest.js:
```diff
@@ -0,0 +1,102 @@
+                },
+            }))
+            .then(() => {
+                // This is a fire and forget response, but one it completes we should be able to verify that we
```
```suggestion
                // This is a fire and forget response, but once it completes we should be able to verify that we
```

tgolen commented on tests/actions/ReportTest.js:
```diff
@@ -0,0 +1,102 @@
+                expect(resultAction).toEqual(REPORT_ACTION);
+            })
+            .then(() => {
+                // Now that we are subscribed we need to simulate a reportComment action Pusher event.
```
> `// Now that we are subscribed`

Subscribed to what? Is this referring to the original Pusher subscription or to Onyx? This reads a little weird because I would expect the thing before this to be code that does subscribing, but the thing before this is adding the report action.

marcaaron replied:
I can make this more explicit.

marcaaron replied:
It probably made more sense before when the thing immediately before this was `subscribeToReportComments()`

tgolen commented on tests/actions/ReportTest.js:
```diff
@@ -0,0 +1,102 @@
+                    reportAction: REPORT_ACTION,
+                });
+
+                // Once this happens we should see the comment get processed by the callback and added to the
```
> `// Once this happens`

"this" could be referring to a number of things, so be more explicit

tgolen replied:
> `the callback`

This is also ambiguous. Is it the pusher callback, the onyx callback, or a different callback entirely?

marcaaron replied:
I can make this more explicit.

tgolen commented on tests/actions/ReportTest.js:
```diff
@@ -0,0 +1,102 @@
+
+                // Verify that our action is no longer in the loading state
+                REPORT_ACTION.loading = false;
+                expect(resultAction).toEqual(REPORT_ACTION);
```
Using a different assertion might make it a little more clear what's being tested and prevents the need to modify the contents of `REPORT_ACTION` (which is not quite intuitive as to why it's necessary to modify `REPORT_ACTION`).

Suggestion:
```
expect(resultAction.loading).toEqual(false);
```

marcaaron replied:
Yeah this is much better

tgolen commented on tests/actions/ReportTest.js:
```diff
@@ -0,0 +1,102 @@
+            })
+            .then(() => {
+                const resultAction = reportActions[ACTION_ID];
+                expect(resultAction).toEqual(REPORT_ACTION);
```
Adding this assertion will help identify that yes, this is definitely an optimistic comment, and not the final result.

```
expect(resultAction.loading).toEqual(true);
```

This, of course, is already implied in the current assertion, but it's an extra mental jump that has to be made which isn't totally obvious.

marcaaron replied:
Yep that makes sense to me. I think it'll be a more interesting test when the `clientID` stuff comes along. But this is mostly what we are looking for now and proves that the action was handled from the Pusher payload.

tgolen replied:
Yeah, this will look cool once the clientID stuff happens, and it will be a perfect use for these tests!

marcaaron committed:
Respond to Tims feedback.

tgolen reviewed (dismissed):
Looks great, thanks for those improvements!

marcaaron committed:
update react-native-onyx

marcaaron committed:
dont test strict equal since the timestamp will make the test flaky

marcaaron committed:
remove momnet

marcaaron commented:
Taking this one off hold .

marcaaron commented:
Also added a README over here if y'all are interested 

https://github.com/Expensify/Expensify.cash/pull/1236

AndrewGable reviewed (dismissed):
Great improvement 👍

marcaaron committed:
fix conflicts

marcaaron committed:
add comment
